cpp/src/io/parquet/reader_impl_helpers.hpp

```diff
@@ -195,4 +215,63 @@ class aggregate_reader_metadata {
                    type_id timestamp_type_id) const;
 };
 
+/**
+ * @brief Converts named columns to index reference columns
+ *
+ */
+class named_to_reference_converter : public ast::detail::expression_transformer {
+ public:
+  named_to_reference_converter(std::optional<std::reference_wrapper<ast::expression const>> expr,
+                               table_metadata const& metadata)
+    : metadata(metadata)
+  {
+    if (!expr.has_value()) return;
+    // create map for column name.
+    std::transform(metadata.schema_info.begin(),
+                   metadata.schema_info.end(),
+                   std::inserter(column_name_to_index, column_name_to_index.end()),
+                   [i = 0](const auto& data) mutable { return make_pair(data.name, i++); });
+    expr.value().get().accept(*this);
+  }
+
+  /**
+   * @copydoc ast::detail::expression_transformer::visit(ast::literal const& )
+   */
+  std::reference_wrapper<ast::expression const> visit(ast::literal const& expr) override;
+  /**
+   * @copydoc ast::detail::expression_transformer::visit(ast::column_reference const& )
+   */
+  std::reference_wrapper<ast::expression const> visit(ast::column_reference const& expr) override;
+  /**
+   * @copydoc ast::detail::expression_transformer::visit(ast::column_name_reference const& )
+   */
+  std::reference_wrapper<ast::expression const> visit(
+    ast::column_name_reference const& expr) override;
+  /**
+   * @copydoc ast::detail::expression_transformer::visit(ast::operation const& )
+   */
+  std::reference_wrapper<ast::expression const> visit(ast::operation const& expr) override;
+
+  /**
+   * @brief Returns the AST to apply on Column chunk statistics.
+   *
+   * @return AST operation expression
+   */
+  [[nodiscard]] std::optional<std::reference_wrapper<ast::expression const>> get_converted_expr()
+    const
+  {
+    return _stats_expr;
+  }
+
+ private:
+  std::vector<std::reference_wrapper<ast::expression const>> visit_operands(
+    std::vector<std::reference_wrapper<ast::expression const>> operands);
+
+  table_metadata const& metadata;
+  std::unordered_map<std::string, size_type> column_name_to_index;
+  std::optional<std::reference_wrapper<ast::expression const>> _stats_expr;
+  std::list<ast::column_reference> _col_ref;
+  std::list<ast::operation> _operators;
```

Comment:
Maybe add a comment about why we choose `std::list` here. I can't remember if I had a comment like this elsewhere in the AST code. In my understanding it's to prevent invalidation of the reference wrappers that would occur if a `std::vector` is resized?

Reply:
That's right. Using `std::list` or `std::deque` to avoid reference invalidation.